crates/aiken-lang/src/tipo/expr.rs

```diff
@@ -976,10 +976,6 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
             kind.is_let(),
         )?;
 
-        if kind.is_expect() && value_typ.contains_opaque() {
-            return Err(Error::ExpectOnOpaqueType { location });
-        }
-
```

Comment:
Move that check to `unify` instead. Since the problem really only occurs when casting from something into an opaque type. So the fix is to simply forbid type-casting on opaque types altogether. This allows to still write things like:

```aiken
expect Some(opaque) = maybe_opaque
```

So long as `opaque` is already fully known. But it will prevent doing something like:

```aiken
expect opaque: MyOpaqueType = data
```

Which is exactly what we want.